R/CallbackTuning.R

```diff
@@ -0,0 +1,135 @@
+#' @title Create Tuning Callback
+#'
+#' @description
+#' Specialized [bbotk::CallbackOptimization] for tuning.
+#' Callbacks allow to customize the behavior of processes in mlr3tuning.
+#' The [callback_tuning()] function creates a [CallbackTuning].
+#' Predefined callbacks are stored in the [dictionary][mlr3misc::Dictionary] [mlr_callbacks] and can be retrieved with [clbk()].
+#' For more information on tuning callbacks see [callback_tuning()].
+#'
+#' @examples
+#' # write archive to disk
+#' callback_tuning("mlr3tuning.backup",
+#'   on_optimization_end = function(callback, context) {
+#'     saveRDS(context$instance$archive, "archive.rds")
+#'   }
+#' )
+CallbackTuning = R6Class("CallbackTuning",
+  inherit = bbotk::CallbackOptimization,
+  public = list(
+
+    #' @field on_eval_after_design (`function()`)\cr
+    #'   Stage called after design is created.
+    #'   Called in `ObjectiveTuning$eval_many()`.
+    on_eval_after_design = NULL,
+
+    #' @field on_eval_after_benchmark (`function()`)\cr
+    #'   Stage called after hyperparameter configurations are evaluated.
+    #'   Called in `ObjectiveTuning$eval_many()`.
+    on_eval_after_benchmark = NULL,
+
+    #' @field on_eval_before_archive (`function()`)\cr
+    #'   Stage called before performance values are written to the archive.
+    #'   Called in `ObjectiveTuning$eval_many()`.
+    on_eval_before_archive = NULL
+  )
+)
+
+#' @title Create Tuning Callback
+#'
+#' @description
+#' Function to create a [CallbackTuning].
+#' Predefined callbacks are stored in the [dictionary][mlr3misc::Dictionary] [mlr_callbacks] and can be retrieved with [clbk()].
+#'
+#' Tuning callbacks can be called from different stages of tuning process.
+#' The stages are prefixed with `on_*`.
+#'
+#' ```
+#' Start Tuning
+#'      - on_optimization_begin
+#'     Start Tuner Batch
+#'          - on_optimizer_before_eval
+#'         Start Evaluation
+#'              - on_eval_after_design
+#'              - on_eval_after_benchmark
+#'              - on_eval_before_archive
+#'         End Evaluation
+#'          - on_optimizer_after_eval
+#'     End Tuner Batch
+#'      - on_result
+#'      - on_optimization_end
+#' End Tuning
+#' ```
+#'
+#' See also the section on parameters for more information on the stages.
+#' A tuning callback works with [bbotk::ContextOptimization] and [ContextEval].
+#'
+#' @details
+#' A callback can write data to its state (`$state`), e.g. settings that affect the callback itself.
+#' Avoid writing large data the state.
```

Comment:
**to** the state

Reply:
Also here I guess it would be informative to know why there are two different contexts (explain the issue with the parallelization).